File: Tests/RoutineProtocolTests.swift
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

import XCTest
@testable import Doigt

/// The known protocols, pinned to the numbers Nuri captured from their sources. Each is
/// checked AFTER `normalized`, because that is what `importRoutine` writes — a band the
/// normalizer demoted or a set it dropped would be a protocol that lands different from
/// its preview.
final class RoutineProtocolTests: XCTestCase {

    private func landed(_ item: RoutineProtocol) -> RoutineDraft { item.draft.normalized }

    private func bands(_ plan: SessionPlan) -> [ClosedRange<Double>?] {
        plan.sets.map(\.targetPercentBand)
    }

    func testEveryProtocolIsSaveableAndStableUnderNormalizing() {
        for item in RoutineProtocol.allCases {
            let draft = landed(item)
            XCTAssertNil(draft.validationIssue, "\(item)")
            XCTAssertEqual(draft.normalized, draft, "\(item) must be normalized already")
            XCTAssertEqual(draft.plan.name, item.title, "\(item)")
            XCTAssertEqual(draft.plan.sets.count, item.draft.plan.sets.count,
                           "\(item): normalizing dropped a set")
        }
    }

    func testEachCallMintsFreshRowIdentity() {
        for item in RoutineProtocol.allCases {
            let a = Set(item.draft.plan.sets.map(\.id))
            let b = Set(item.draft.plan.sets.map(\.id))
            XCTAssertTrue(a.isDisjoint(with: b), "\(item)")
        }
    }

    func testDailyNoHangs() {
        let plan = landed(.dailyNoHangs).plan
        XCTAssertEqual(plan.handMode, .alternateEachRep)
        XCTAssertEqual([plan.holdSeconds, plan.restSeconds, plan.setBreakSeconds], [10, 10, 10],
                       "alternating: a 10 s gap + the other hand's 10 s pull = 20 s a hand")
        XCTAssertEqual(plan.sets.map(\.grip), [
            GripSpec(edgeMM: 20, fingers: .four,       position: .halfCrimp),
            GripSpec(edgeMM: 20, fingers: .frontThree, position: .drag),
            GripSpec(edgeMM: 20, fingers: .frontTwo,   position: .drag),
            GripSpec(edgeMM: 20, fingers: .middleTwo,  position: .drag),
            GripSpec(edgeMM: 20, fingers: .frontTwo,   position: .halfCrimp),
            GripSpec(edgeMM: 20, fingers: .middleTwo,  position: .halfCrimp),
        ])
        XCTAssertEqual(plan.sets.map(\.repsPerSide), [6, 6, 2, 2, 2, 2])
        XCTAssertEqual(bands(plan), Array(repeating: 0.35...0.45, count: 6))
        XCTAssertEqual(PlanMath.totalReps(plan), 40)
        XCTAssertEqual(landed(.dailyNoHangs).sessionsPerDay, 2)
    }

    func testC4WarmUp() {
        let draft = landed(.c4WarmUp)
        let plan = draft.plan
        XCTAssertEqual(plan.handMode, .alternateEachSet)
        XCTAssertEqual([plan.holdSeconds, plan.restSeconds, plan.setBreakSeconds], [5, 10, 30])
        XCTAssertEqual(plan.sets.map(\.repsPerSide), Array(repeating: 2, count: 6))
        XCTAssertEqual(bands(plan), [0.45...0.55, 0.65...0.75, 0.75...0.85,
                                     0.85...0.95, 0.55...0.65, 0.75...0.85])
        XCTAssertTrue(draft.isOnDemand)
    }

    /// A band that gated would pause the clock on a pull above the max on file.
    func testC4MaxDrawsItsBandButNeverGatesOnIt() {
        let draft = landed(.c4Max)
        let plan = draft.plan
        XCTAssertEqual([plan.holdSeconds, plan.restSeconds, plan.setBreakSeconds], [4, 10, 60])
        XCTAssertEqual(plan.sets.map(\.repsPerSide), [3, 3, 3])
        XCTAssertEqual(bands(plan), Array(repeating: 0.80...1.0, count: 3))
        XCTAssertFalse(plan.pausesOutsideTargetBand)
        XCTAssertTrue(draft.isOnDemand)
    }

    /// The ceiling is the prescription, so it gates; hands alternate every pull.
    func testFingerRehabGatesAtItsCeilingAndAlternates() {
        let plan = landed(.fingerRehab).plan
        XCTAssertEqual(plan.handMode, .alternateEachRep)
        XCTAssertEqual([plan.holdSeconds, plan.restSeconds, plan.setBreakSeconds], [10, 50, 50],
                       "alternating: a 50 s gap + the other hand's 10 s pull = a minute a hand")
        XCTAssertEqual(plan.sets.map(\.repsPerSide), Array(repeating: 5, count: 5))
        XCTAssertEqual(bands(plan), Array(repeating: 0.15...0.25, count: 5))
        XCTAssertTrue(plan.pausesOutsideTargetBand)
        XCTAssertNotNil(RoutineProtocol.fingerRehab.caution)
    }
}

File: android/engine/src/test/kotlin/run/nuri/getagrip/engine/RoutineProtocolTests.kt
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

package run.nuri.getagrip.engine

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

/// The known protocols, pinned to the numbers Nuri captured from their sources. Each is
/// checked AFTER `normalized`, because that is what `importRoutine` writes — a band the
/// normalizer demoted or a set it dropped would be a protocol that lands different from its
/// preview. Twin of Tests/RoutineProtocolTests.swift.
class RoutineProtocolTests {
    private fun landed(item: RoutineProtocol): RoutineDraft = item.draft.normalized
    private fun bands(plan: SessionPlan) = plan.sets.map { it.targetPercentBand }
    private fun timing(plan: SessionPlan) = listOf(plan.holdSeconds, plan.restSeconds, plan.setBreakSeconds)

    @Test fun everyProtocolIsSaveableAndStableUnderNormalizing() {
        for (item in RoutineProtocol.allCases) {
            val draft = landed(item)
            assertNull(draft.validationIssue, "$item")
            assertEquals(draft, draft.normalized, "$item must be normalized already")
            assertEquals(item.title, draft.plan.name, "$item")
            assertEquals(item.draft.plan.sets.size, draft.plan.sets.size, "$item: normalizing dropped a set")
        }
        assertEquals(listOf(RoutineProtocol.dailyNoHangs, RoutineProtocol.c4WarmUp, RoutineProtocol.c4Max,
            RoutineProtocol.fingerRehab), RoutineProtocol.allCases)
    }

    @Test fun eachCallMintsFreshRowIdentity() {
        for (item in RoutineProtocol.allCases) {
            val a = item.draft.plan.sets.map { it.id }.toSet()
            val b = item.draft.plan.sets.map { it.id }.toSet()
            assertTrue(a.intersect(b).isEmpty(), "$item")
        }
    }

    @Test fun dailyNoHangs() {
        val plan = landed(RoutineProtocol.dailyNoHangs).plan
        assertEquals(HandMode.alternateEachRep, plan.handMode)
        assertEquals(listOf(10, 10, 10), timing(plan),
            "alternating: a 10 s gap + the other hand's 10 s pull = 20 s a hand")
        assertEquals(listOf(
            GripSpec(20, FingerSet.four, GripPosition.halfCrimp),
            GripSpec(20, FingerSet.frontThree, GripPosition.drag),
            GripSpec(20, FingerSet.frontTwo, GripPosition.drag),
            GripSpec(20, FingerSet.middleTwo, GripPosition.drag),
            GripSpec(20, FingerSet.frontTwo, GripPosition.halfCrimp),
            GripSpec(20, FingerSet.middleTwo, GripPosition.halfCrimp),
        ).map { it.key }, plan.sets.map { it.grip.key })
        assertEquals(listOf(6, 6, 2, 2, 2, 2), plan.sets.map { it.repsPerSide })
        assertEquals(List(6) { 0.35..0.45 }, bands(plan))
        assertEquals(40, PlanMath.totalReps(plan))
        assertEquals(2, landed(RoutineProtocol.dailyNoHangs).sessionsPerDay)
    }

    @Test fun c4WarmUp() {
        val draft = landed(RoutineProtocol.c4WarmUp)
        val plan = draft.plan
        assertEquals(HandMode.alternateEachSet, plan.handMode)
        assertEquals(listOf(5, 10, 30), timing(plan))
        assertEquals(List(6) { 2 }, plan.sets.map { it.repsPerSide })
        assertEquals(listOf(0.45..0.55, 0.65..0.75, 0.75..0.85, 0.85..0.95, 0.55..0.65, 0.75..0.85), bands(plan))
        assertTrue(draft.isOnDemand)
    }

    /// A band that gated would pause the clock on a pull above the max on file.
    @Test fun c4MaxDrawsItsBandButNeverGatesOnIt() {
        val draft = landed(RoutineProtocol.c4Max)
        val plan = draft.plan
        assertEquals(listOf(4, 10, 60), timing(plan))
        assertEquals(listOf(3, 3, 3), plan.sets.map { it.repsPerSide })
        assertEquals(List(3) { 0.80..1.0 }, bands(plan))
        assertFalse(plan.pausesOutsideTargetBand)
        assertTrue(draft.isOnDemand)
    }

    /// The ceiling is the prescription, so it gates; hands alternate every pull.
    @Test fun fingerRehabGatesAtItsCeilingAndAlternates() {
        val plan = landed(RoutineProtocol.fingerRehab).plan
        assertEquals(HandMode.alternateEachRep, plan.handMode)
        assertEquals(listOf(10, 50, 50), timing(plan),
            "alternating: a 50 s gap + the other hand's 10 s pull = a minute a hand")
        assertEquals(List(5) { 5 }, plan.sets.map { it.repsPerSide })
        assertEquals(List(5) { 0.15..0.25 }, bands(plan))
        assertTrue(plan.pausesOutsideTargetBand)
        assertNotNull(RoutineProtocol.fingerRehab.caution)
    }
}
